tests/core/abi-utils/test_abi_utils.py

```diff
@@ -173,3 +694,440 @@ def test_event_signature_to_log_topic(event_signature, expected):
+            "t": {"x": 11, "y": 12},
+            "a": 13,
+            "b": [[(14, 15), (16, 17)], [{"x": 18, "y": 19}]],
+        },
+        GET_ABI_INPUTS_OUTPUT,
+    ),
+    (
+        TEST_FUNCTION_ABI,
+        {
+            "s": {"a": 1, "b": [2, 3, 4], "c": [(5, 6), (7, 8), (9, 10)]},
+            "t": (11, 12),
+            "a": 13,
+            "b": [[(14, 15), (16, 17)], [(18, 19)]],
+        },
+        GET_ABI_INPUTS_OUTPUT,
+    ),
+    (
+        TEST_FUNCTION_ABI,
+        {
+            "s": (1, [2, 3, 4], [(5, 6), (7, 8), (9, 10)]),
+            "t": (11, 12),
+            "a": 13,
+            "b": [[(14, 15), (16, 17)], [(18, 19)]],
+        },
+        GET_ABI_INPUTS_OUTPUT,
+    ),
+    (
+        TEST_FUNCTION_ABI,
+        (
+            (1, [2, 3, 4], [(5, 6), (7, 8), (9, 10)]),
+            (11, 12),
+            13,
+            [[(14, 15), (16, 17)], [(18, 19)]],
+        ),
+        GET_ABI_INPUTS_OUTPUT,
+    ),
+    (
+        TEST_FUNCTION_ABI,
+        {
+            "s": {"a": 1, "b": [2, 3, 4], "c": [(5, 6), (7, 8), MyXYTuple(x=9, y=10)]},
+            "t": MyXYTuple(x=11, y=12),
+            "a": 13,
+            "b": [
+                [MyXYTuple(x=14, y=15), MyXYTuple(x=16, y=17)],
+                [MyXYTuple(x=18, y=19)],
+            ],
+        },
+        GET_ABI_INPUTS_OUTPUT,
+    ),
+)
+
+
+@pytest.mark.parametrize(
+    "abi, args, expected",
+    GET_ABI_INPUTS_TESTS,
+)
+def test_get_aligned_abi_inputs(abi, args, expected):
+    assert get_aligned_abi_inputs(abi, args) == expected
+
+
+@pytest.mark.parametrize(
+    "abi, args, error",
+    (
+        (
+            # raise TypeError, expects list of tuples for input 's.c'
+            TEST_FUNCTION_ABI,
+            {
+                "s": {"a": 1, "b": [2, 3, 4], "c": ["56", (7, 8), (9, 10)]},
+                "t": (11, 12),
+                "a": 13,
+                "b": [[(14, 15), (16, 17)], [(18, 19)]],
+            },
+            'Expected non-string sequence for "tuple" component type: got 56',
+        ),
+        (
+            # raise TypeError, expects list of tuples for input 's.c'
+            TEST_FUNCTION_ABI,
+            {
+                "s": {"a": 1, "b": [2, 3, 4], "c": {(5, 6), (7, 8), (9, 10)}},
+                "t": (11, 12),
+                "a": 13,
+                "b": [[(14, 15), (16, 17)], [(18, 19)]],
+            },
+            'Expected non-string sequence for "tuple[]" component type: got '
+            "{(9, 10), (5, 6), (7, 8)}",
+        ),
+    ),
+)
+def test_get_aligned_abi_inputs_raises_type_error_for_incorrect_input_types(
+    abi, args, error
+):
+    with pytest.raises(
+        TypeError,
+        match=re.escape(error),
+    ):
+        get_aligned_abi_inputs(abi, args)
+
+
+@pytest.mark.parametrize("type", (("fallback"), ("receive")))
+def test_get_aligned_abi_inputs_raises_value_error_for_fallback_or_receive_types(type):
```

Comment:
These last two tests look exactly the same except for a small part of the error message. I think we can parametrize with ids here as well if we wanted. Take or leave.